### PR TITLE
Download files instead of inlining into HTML

### PR DESCRIPTION
**Pull request by nwalters512**
This PR aims to reduce server load when students are uploading files that are many megabytes large.

## Background

File uploading was kinda hacked into PL back in the day by yours truly. It worked via base-64-encoded files in the submitted answer. This worked fine when the only thing folks were uploading were small source files (think 10s of kilobytes at most).

However, since then, folks have started uploading very large files, e.g. 5MB Jupyter notebooks with embedded images, or 20MB PDFs. When that data is base-64-encoded and inlined into HTML, it presents a real burden for our HTML-processing machinery. Specifically, getting that data out of the Python process and then running it through an HTML parser takes several seconds for large files, of which a good deal of that is actually *synchronous* parsing that blocks the event loop.

## Improvements

This PR adds a new API route for downloading a file associated with a submission. So, instead of inlining the file's contents in the HTML, we can just pass along a name and rely on the client to lazily fetch the files when they are needed (e.g. to show a preview or to download the file).

It also introduces changes to the elements that are most often used with very large files: `pl-file-upload` and `pl-file-preview`. These elements will use the above API to fetch file contents.

### CODE REVIEW

**Comment by echuber2** [comment]
I don't know if it's possible to get away from base64 completely, but it also bloats the size by ~30%.

**Comment by nwalters512** [comment]
@echuber2 we'll most likely do that eventually via an overhaul of the way that files are handled. That's a much larger change though, and something we'll have to be very careful about to avoid breaking existing questions. I wouldn't expect to see any movement there for a while.

File: elements/pl-file-upload/pl-file-upload.mustache
Comment (eliotwrobson) [inline elements/pl-file-upload/pl-file-upload.mustache:5]: Is the `has_submitted_files` key necessary? Can't you just case on whether `submitted_file_names` is empty?

File: elements/pl-file-upload/pl-file-upload.py
Comment (eliotwrobson) [inline elements/pl-file-upload/pl-file-upload.py:85]:
```suggestion
        return chevron.render(f, html_params).strip()
```

File: elements/pl-file-preview/pl-file-preview.py
Comment (eliotwrobson) [inline elements/pl-file-preview/pl-file-preview.py:34]: Files can now be assembled in one line (and possibly some of the logic above can be simplified?)
```suggestion
        files = [
            {
                    "name": file["name"],
                    "index": idx,
            }
            for idx, file in enumerate(submitted_files)
        ]
```

**Review by eliotwrobson — COMMENTED** [review]
Is this an improvement on https://github.com/PrairieLearn/PrairieLearn/pull/6988? If so, I think that PR can be closed.

As I mentioned on that PR, this is a good opportunity to add type hints + checking to these Python files.

File: pages/submissionFiles/submissionFiles.sql
Comment (mwest1066) [inline pages/submissionFiles/submissionFiles.sql:16]: Elsewhere we've used the pattern that `$instance_question_id` is passed as NULL to indicate not to check it, which would then give
```suggestion
        AND (($instance_question_id::bigint IS NULL) OR (v.instance_question_id = $instance_question_id))
```

Comment (mwest1066) [reply pages/submissionFiles/submissionFiles.sql:16]: (but I'm also fine with the explicit boolean flag if you prefer that)

Comment (nwalters512) [reply pages/submissionFiles/submissionFiles.sql:16]: I like that better! I copied this from #5500 which git-blames back to me 🙃

File: elements/pl-file-upload/pl-file-upload.mustache
Comment (nwalters512) [reply elements/pl-file-upload/pl-file-upload.mustache:5]: I sure can! This was based on existing code.

File: pages/submissionFiles/submissionFiles.sql
Comment (nwalters512) [reply pages/submissionFiles/submissionFiles.sql]: Changed in 3aa81c1.

File: elements/pl-file-upload/pl-file-upload.py
Comment (eliotwrobson) [inline elements/pl-file-upload/pl-file-upload.py:32]: Functions that don't return anything need a `None` return annotation. Otherwise, the typechecker assumes the return type to be `Any` (in other words, doesn't enforce any behavior).
```suggestion
def add_format_error(data: pl.QuestionData, error_string: str) -> None:
```

File: elements/pl-file-upload/pl-file-upload.py
Comment (eliotwrobson) [inline elements/pl-file-upload/pl-file-upload.py:75]:
```suggestion
    submitted_file_names = {x.get("name") for x in files} & set(file_names)
```

File: elements/pl-file-upload/pl-file-upload.py
Comment (eliotwrobson) [inline elements/pl-file-upload/pl-file-upload.py:103]: Probably want to catch if anything happens here (even though it will be a `ValueError` most of the time).
```suggestion
    except Exception:
```

File: elements/pl-file-upload/pl-file-upload.py
Comment (eliotwrobson) [inline elements/pl-file-upload/pl-file-upload.py:50]:
```suggestion
def render(element_html: str, data: pl.QuestionData) -> str:
```

File: elements/pl-file-preview/pl-file-preview.py
Comment (eliotwrobson) [inline elements/pl-file-preview/pl-file-preview.py:15]: No need to make this a variable if used once.
```suggestion
    html_params["required_files"] = data["params"].get("_required_file_names", [])
```

File: elements/pl-file-preview/pl-file-preview.py
Comment (eliotwrobson) [inline elements/pl-file-preview/pl-file-preview.py:8]: Shouldn't we keep this `check_attribs` call to enforce there are no attributes?

Comment (nwalters512) [reply elements/pl-file-preview/pl-file-preview.py:8]: I suppose, though I don't think we really care if there are attributes or not.

File: exampleCourse/questions/demo/autograder/codeUpload/question.html
Comment (nwalters512) [inline exampleCourse/questions/demo/autograder/codeUpload/question.html:20]: TODO: revert before merging.

File: lib/question.js
Comment (nwalters512) [inline lib/question.js:1396]:
```suggestion
```

File: pages/submissionFiles/submissionFiles.js
Comment (nwalters512) [inline pages/submissionFiles/submissionFiles.js:67]: TODO: delete before merging.

File: pages/submissionFiles/submissionFiles.sql
Comment (nwalters512) [inline pages/submissionFiles/submissionFiles.sql:12]:
```suggestion
        -- this submission is actually accessible by the user. Both will have been
```

File: elements/pl-file-preview/pl-file-preview.js
Comment (nwalters512) [inline elements/pl-file-preview/pl-file-preview.js:75]: The error experience is kind of janky; here's hoping it doesn't happen much!

File: elements/pl-file-upload/pl-file-upload.js
Comment (nwalters512) [inline elements/pl-file-upload/pl-file-upload.js:30]: I refactored this file into a class, which unfortunately made this difficult to review. I can refactor the existing behavior in another PR to make the diff cleaner if folks would like that.

Comment (mwest1066) [reply elements/pl-file-upload/pl-file-upload.js:12]: I think we can probably just test the functionality rather than needing a careful diff? Or are there particular code changes in this file that you'd like eyes on?

Comment (nwalters512) [reply elements/pl-file-upload/pl-file-upload.js:12]: I only propose this so that I/we could reason about what actually _did_ change - the diff right now has a lot of false positives.

Comment (nwalters512) [reply elements/pl-file-upload/pl-file-upload.js:12]: Good call - that definitely helped a bunch!

File: server.js
Comment (nwalters512) [inline server.js:934]: I think `.../file` might be more idiomatic than `/files`? We tend to use singular nouns in most of our routes.

File: lib/question.js
Comment (mwest1066) [inline lib/question.js:1381]: I think we also need this in the `else` branch of the `if (!assessment)` statement so that student questions will also have this defined. At the moment we get this on student pages:
image

File: elements/pl-file-preview/pl-file-preview.py
Comment (mwest1066) [reply elements/pl-file-preview/pl-file-preview.py:8]: I'd prefer to enforce strict compliance here (don't allow attributes if none are permitted) as a matter of policy. Being strict now can only help us with backwards compatibility in the future.

File: elements/pl-file-preview/pl-file-preview.js
Comment (nwalters512) [reply elements/pl-file-preview/pl-file-preview.js:32]: I don't believe this is a regression; `pl-file-upload` displays files in the order that they appear in the `file-names` attribute, whereas `pl-file-preview` displays the files in the order in which they appear in `params._files`. Perhaps we could update the code to display the `pl-file-preview` files in the order in which they appear in `params._required_file_names`?

File: lib/question.js
Comment (nwalters512) [reply lib/question.js:1381]: Good call! Will add.

File: elements/pl-file-upload/pl-file-upload.css
Comment (nwalters512) [reply elements/pl-file-upload/pl-file-upload.css:1]: Good catch, I'll look into this.

**Comment by nwalters512** [comment]
Apologies for the codeowners tag, I messed up a rebase and it appears as though GitHub doesn't remove the review requests after fixing that 🤦

File: lib/question.js
Comment (nwalters512) [reply lib/question.js:1381]: Fixed in af78b02.

File: server.js
Comment (nwalters512) [reply server.js]: Changed to `/file` in ac775c4.

File: elements/pl-file-preview/pl-file-preview.py
Comment (nwalters512) [reply elements/pl-file-preview/pl-file-preview.py]: I added attribute checking back in fa7326a.

File: elements/pl-file-upload/pl-file-upload.css
Comment (nwalters512) [reply elements/pl-file-upload/pl-file-upload.css:1]: Resolved in 3ce14fd - turns out that `dz-message` class was important!

File: elements/pl-file-preview/pl-file-preview.js
Comment (nwalters512) [reply elements/pl-file-preview/pl-file-preview.js:32]: This was pretty easy to implement: 7174621. I wrote some tests for this too.

File: elements/pl-file-preview/pl_file_preview_utils.py
Comment (eliotwrobson) [inline elements/pl-file-preview/pl_file_preview_utils.py:5]: Lowercase dict annotations ara python 3.10 only feature! Also missing type of this dict.

File: elements/pl-file-preview/test_pl_file_preview_utils.py
Comment (eliotwrobson) [inline elements/pl-file-preview/test_pl_file_preview_utils.py:4]: Missing annotations

**Review by eliotwrobson — DISMISSED** [review]
See comments.

File: elements/pl-file-preview/pl-file-preview.js
Comment (mwest1066) [reply elements/pl-file-preview/pl-file-preview.js:32]: Nice!

Comment (jonatanschroeder) [reply elements/pl-file-preview/pl-file-preview.js:32]: I see that files that are not required are not sorted in any meaningful order. Should they be? I'm thinking of workspaces with wildcards, which don't appear as required files.

Comment (mwest1066) [reply elements/pl-file-preview/pl-file-preview.js:32]: Good point. Would it be cleaner to always alphabetize?

Comment (nwalters512) [reply elements/pl-file-preview/pl-file-preview.js:32]: Good point - plain old alphabetization is almost certainly better. I'll make that change.

File: elements/pl-file-preview/pl-file-preview.js
Comment (mwest1066) [inline elements/pl-file-preview/pl-file-preview.js:90]: In this case can we write this text not inside the code element? For example, in the files below it looks like `foo.pdf` has rendered some content until you read the text. I'm thinking it might be better to just have plain text for the message.

image

File: elements/pl-file-preview/pl-file-preview.js
Comment (mwest1066) [inline elements/pl-file-preview/pl-file-preview.js:98]: `showErrorMessage()` doesn't seem to take an argument?

**Review by mwest1066 — APPROVED** [review]

File: elements/pl-file-preview/pl-file-preview.js
Comment (nwalters512) [reply elements/pl-file-preview/pl-file-preview.js:32]: Alphabetization was implemented in 754f260.

File: elements/pl-file-preview/pl-file-preview.js
Comment (nwalters512) [reply elements/pl-file-preview/pl-file-preview.js:98]: Good catch, will remove.

Comment (nwalters512) [reply elements/pl-file-preview/pl-file-preview.js:98]: Actually, I might repurpose the error display for the "Content preview is not available for this type of file" message, in which case I'll actually change `showErrorMessage` to accept a string.

File: elements/pl-file-preview/pl-file-preview.js
Comment (nwalters512) [reply elements/pl-file-preview/pl-file-preview.js]: Yep - I tweaked this in 19a115b. How does this look?

Screenshot 2023-02-07 at 15 28 52

File: exampleCourse/questions/demo/autograder/codeUpload/question.html
Comment (nwalters512) [reply exampleCourse/questions/demo/autograder/codeUpload/question.html:20]:
```suggestion
  <pl-file-upload file-names="fib.py"></pl-file-upload>
```

File: elements/pl-file-preview/pl-file-preview.mustache
Comment (nwalters512) [reply elements/pl-file-preview/pl-file-preview.mustache:23]: Hmm, I can't reproduce that with a text file submission on the latest commit to this branch. Is there a particular file type (e.g. image or binary file) that you tested this with?

Comment (nwalters512) [reply elements/pl-file-preview/pl-file-preview.mustache:23]: Chatted with Matt offline; he was unable to reproduce this.

**Review by mwest1066 — APPROVED** [review]

**Pull request merged**
